executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,391 @@
+                        rhs.map_variables(&mut map_var_idx).to_rpn(),
+                        *expected_equal,
+                    ));
+                }
+                Effect::MachineCall(id, arguments) => {
+                    let mut result_vars = vec![];
+
+                    arguments.iter().for_each(|a| match a {
+                        MachineCallArgument::Unknown(v) => {
+                            let idx = map_var_idx(v);
+                            result_vars.push(idx);
+                        }
+                        MachineCallArgument::Known(_) => {}
+                    });
+
+                    actions.push(InterpreterAction::MachineCall(
+                        result_vars,
+                        *id,
+                        arguments
+                            .iter()
+                            .map(|a| match a {
+                                MachineCallArgument::Unknown(_) => MachineCallArgument::Unknown(0),
+                                MachineCallArgument::Known(v) => {
+                                    MachineCallArgument::Known(v.map_variables(&mut map_var_idx))
+                                }
+                            })
+                            .collect(),
+                    ));
+                }
+            }
+        }
+
+        // write variables back
+        let vars_known: Vec<_> = effects.iter().flat_map(written_vars_in_effect).collect();
+        vars_known.iter().for_each(|var| {
+            match var {
+                Variable::Cell(cell) => {
+                    let idx = var_idx.get(var).unwrap();
+                    actions.push(InterpreterAction::WriteCell(*idx, cell.clone()));
+                    actions.push(InterpreterAction::WriteKnown(cell.clone()));
+                }
+                Variable::Param(i) => {
+                    let idx = var_idx.get(var).unwrap();
+                    actions.push(InterpreterAction::WriteParam(*idx, *i));
+                }
+                Variable::MachineCallReturnValue(_) => {
+                    // This is just an internal variable.
+                }
+            }
+        });
+
+        Self {
+            first_column_id,
+            column_count,
+            vars,
+            actions,
+        }
+    }
+
+    // Execute the machine effects for the given the parameters
+    pub fn call<Q: QueryCallback<T>>(&self, params: WitgenFunctionParams<'_, T>) {
+        let known = known_to_slice(self.column_count, params.known, params.data.len);
+        let data = params.data.into_mut_slice();
+        let pparams = params.params.into_mut_slice();
+
+        let mut vars = vec![None; self.vars];
+
+        let mut eval_stack = vec![];
+        for action in &self.actions {
+            match action {
+                InterpreterAction::AssignExpression(idx, e) => {
+                    let val = self.evaluate_expression(&mut eval_stack, &vars, e);
+                    assert!(vars[*idx].replace(val).is_none());
```

Comment:
Instead of this runtime assertion, we can actually check in the conversion inside the `new` function if each variable is written to exactly once (and only read after it has been written).